### DISCUSSION TITLE
Why aren't the IR classes public?

### DISCUSSION THREAD

**Discussion by bricelam** [microsoft/Power-Fx#2710] [Q&A] [answered]
So many interesting extensions and applications of Power Fx (e.g. the Interpreter, Dataverse, OpenAPI/OData Connectors) seem to depend on IRResult, IRNodeVisitor, and the IntermediateNode classes. Why aren't these types public, and are there plans to make them public?

Until then, are we expected to work directly with CheckResult, TexlVisitor, and the TexlNode classes instead? This will require us to re-implement binding, type inference, etc. ourselves.

Note, I'm primarily interested in translating expressions, not evaluating them.

**Comment by MikeStall** [comment, marked as answer]
1. We do eventually want to make the IR classes public.   The challenge has been a)_low demand, b) once they're public, we can't change them - so we need to be very confident they're the right shape. Per that, we generally keep things internal as much as we can, and then make public surface area in a very controlled, scrutinized way. 
2. CheckResult will handle binding, infererence, etc. 


For example, DType is the internal type class. But not a good, stable, public surface. So we wrapped that in FormulaType - and make that public. 

Can you tell me more about what kind of translation you want?

**Reply by bricelam** [reply]
I'm an enterprise developer. We use [EF Core](https://github.com/dotnet/efcore), and we want to allow our users to specify some filters, aggregate operations, and per-row operations (projections with expressions). Our users our very familiar with Excel, so instead of inventing and implementing our own expression syntax, I'm looking at whether or not it's feasible to reuse Power Fx.

I'm still figuring out exactly what it will look like since dynamic models are difficult to work with in EF, but ultimately, I'd like to translate to LINQ [Expression](https://learn.microsoft.com/en-us/dotnet/api/system.linq.expressions.expression?view=net-8.0) trees. Here are some off-the-cuff thoughts of what the API might look like:

```cs
var aPosts = dbContext.FromPowerFx(
    """
    Filter(Blogs, StartsWith(Title, "A"))
    """);

var aPosts2 = dbContext.Blogs.WherePowerFx(
   """
   StartsWith(Title, "A")
   """);

var posts = dbContext.Blogs.SelectPowerFx(
   """
   Id: =Id
   Title: =Title
   IsAPost: =StartsWith(Title, "A")
   """);
```

**Reply by bricelam** [reply]
This is finally starting to take shape over in [bricelam/Bricelam.PowerFx.Linq](https://github.com/bricelam/Bricelam.PowerFx.Linq)

**Reply by lesaltzm** [reply]
It looks like that repo is private, but that's an exciting use case! We'd love to see it when you're ready to make it public.

**Reply by bricelam** [reply]
Sorry about that! I was a bit too hasty. It's public now.
